### PR TITLE
Add proxy options.

### PR DESCRIPTION
**Pull request by twitchax**
Allow the developer to set proxy options.

```csharp
public class MyController : Controller
{
    [Route("api/posts/{postId}")]
    public Task GetPosts(int postId)
    {
        var options = ProxyOptions.Instance
            .WithShouldAddForwardedHeaders(false)
            .WithBeforeSend((c, hrm) =>
            {
                // Set something that is needed for the downstream endpoint.
                hrm.Headers.Authorization = new AuthenticationHeaderValue("Basic");
            })
            .WithAfterReceive((c, hrm) =>
            {
                // Alter the conent in  some way before sending back to client.
                var newContent = new StringContent("It's all greek...er, Latin...to me!");
                hrm.Content = newContent;
            })
            .WithHandleFailure((c, e) =>
            {
                // Return a custom error response.
                c.Response.StatusCode = 403;
                c.Response.WriteAsync("Things borked.");
            });

        return this.ProxyAsync($"https://jsonplaceholder.typicode.com/posts/{postId}");
    }
}
```

Also, adds `X-Forwarded-*` and `Forwarded` headers.

### CODE REVIEW

**Comment by twitchax** [comment]
@digitalpowers, @wasabii, @urffin, take a look?

**Comment by wasabii** [comment]
Neat. I believe the host header should probably have the port in it.

And I think you need to do special formatting for ipv6.

**Comment by urffin** [comment]
Seems not bad, do you think about `Func<>` instead `Action` in callback, so instead 

```
.WithAfterReceive((c, hrm) =>
    {
        // Alter the content in  some way before sending back to client.
        var newContent = new StringContent("It's all greek...er, Latin...to me!");
        hrm.Content = newContent;
    })
```

use

```
.WithAfterReceive((c, hrm) =>
    {
        // Alter the content in  some way before sending back to client.
        return  new StringContent("It's all greek...er, Latin...to me!");
    })
```

**Comment by urffin** [comment]
@twitchax, i can look deeper only in saturday

File: README.md [hunk 1]
```diff
@@ -56,6 +56,39 @@ public class MyController : Controller
...
+    public Task GetPosts(int postId)
+    {
+        var options = ProxyOptions.Instance
+            .WithShouldAddForwardedHeaders(false)
+            .WithBeforeSend((c, hrm) =>
+            {
+                // Set something that is needed for the downstream endpoint.
+                hrm.Headers.Authorization = new AuthenticationHeaderValue("Basic");
+            })
+            .WithAfterReceive((c, hrm) =>
```
Comment (urffin) [inline README.md]: I understad what's wrong with this: you can't _partially change_ Content, you can just fully replace it. But for partially change you shoulr read Content, and for do this you should use async methods, so Action seems a bit not correct type for callback, it's more like `Func<context, response, Task>`

File: src/Core/Helpers.cs [hunk 2]
```diff
@@ -28,39 +30,44 @@ internal static IEnumerable<Assembly> GetReferencingAssemblies()
...
                 {
                     // If the failures are not caught, then write a generic response.
-                    context.Response.StatusCode = 500;
+                    context.Response.StatusCode = 502;
                     await context.Response.WriteAsync($"Request could not be proxied.\n\n{e.Message}\n\n{e.StackTrace}.").ConfigureAwait(false);
                     return;
                 }
 
-                await onFailure(context, e).ConfigureAwait(false);
+                options?.HandleFailure?.Invoke(context, e);
```
Comment (urffin) [inline src/Core/Helpers.cs]: not needed Elvis operator here, we already check, that `options`,  and `HandleFailure` not null

Comment (twitchax) [reply src/Core/Helpers.cs]: Ahh, good catch...will fix.

File: README.md (hunk 1 above)
Comment (twitchax) [reply README.md]: Yeah, I have been going back and forth on this.  I think the scenario you mention is compelling, so I will change this.

**Comment by twitchax** [comment]
@urffin, addressed your comments.

**Comment by urffin** [comment]
@twitchax, don't forget merge to master :-)

**Pull request merged**